Self-Hosted Runners, EKS, and Leapp Design Decisions (#684)
cloudposse/docs: docs/layers/github-actions/design-decisions/decide-on-self-hosted-runner-architecture.mdx

---
title: "Decide on Self-Hosted Runner Architecture"
sidebar_label: Runner Architecture
description: Decide how to create self-hosted runners
---

import Intro from "@site/src/components/Intro";
import Note from '@site/src/components/Note';

<Intro>

> Decide on how to operate self-hosted runners that are used to run GitHub Actions workflows. These runners can be set up in various ways and allow us to avoid platform fees while running CI jobs in private infrastructure, enabling access to VPC resources. This approach is ideal for private repositories, providing control over instance size, architecture, and control costs by leveraging spot instances. The right choice depends on your platform, whether you’re using predominantly EKS, ECS, or Lambda.”

Reviewer: ```suggestion
Decide on how to operate self-hosted runners that are used to run GitHub Actions workflows. These runners can be set up in various ways and allow us to avoid platform fees while running CI jobs in private infrastructure, enabling access to VPC resources. This approach is ideal for private repositories, providing control over instance size, architecture, and control costs by leveraging spot instances. The right choice depends on your platform, whether you’re using predominantly EKS, ECS, or Lambda.
```

</Intro>

## Problem

> When using GitHub Actions, you have the option of both GitHub Cloud-hosted and self-hosted runners, which can be used together. However, there are cases where self-hosted runners are essential. For example, if you’re using GitOps, private runners are needed to access resources inside a VPC, such as databases, Kubernetes API endpoints, and Kafka servers.
> 
> While self-hosted runners work well for private infrastructure, they’re not suitable for public or open-source repositories, as they can expose sensitive resources. If your organization contributes to open source, this must be considered when deciding on a runner strategy.
> 
> Your approach to self-hosted runners depends on your infrastructure. If you use Kubernetes, it’s often best to run your runners on Kubernetes. If you rely on ECS or Lambdas, you’ll want to avoid unnecessary Kubernetes dependencies and use a different hosting method for your runners.
> 
> For Kubernetes-based setups, proper node pool configuration with Karpenter is key to maintaining stability and ensuring effective auto-scaling with spot and on-demand instances. We offer several options for deploying self-hosted runners, including EKS, Philips Labs’ solution, or using an Auto Scaling Group (ASG). This decision focuses on the type, placement and management of the runners.

Reviewer: ```suggestion
When using GitHub Actions, you can opt for both GitHub Cloud-hosted and self-hosted runners, and they can complement each other. In some cases, self-hosted runners are essential—particularly for accessing resources within a VPC, such as databases, Kubernetes API endpoints, or Kafka servers, which is common in GitOps workflows.

However, while self-hosted runners are ideal for private infrastructure, they pose risks in public or open-source repositories due to potential exposure of sensitive resources. If your organization maintains open-source projects, this should be a critical consideration, and we recommend using cloud-hosted runners for those tasks.

The hosting approach for self-hosted runners should align with your infrastructure. If you use Kubernetes, it's generally best to run your runners on Kubernetes. Conversely, if your infrastructure relies on ECS or Lambdas, you may want to avoid unnecessary Kubernetes dependencies and opt for alternative hosting methods.

In Kubernetes-based setups, configuring node pools with Karpenter is key to maintaining stability and ensuring effective auto-scaling with a mix of spot and on-demand instances. However, tuning this setup can be challenging, especially with recent changes to ARC, where the [newer version does not support multiple labels for runner groups](https://github.com/actions/actions-runner-controller/issues/2445), leading to community disagreement over trade-offs. We provide multiple deployment options for self-hosted runners, including EKS, Philips Labs' solution, and Auto Scaling Groups (ASG), tailored to your specific runner management needs.
```

---
title: "Decide on Self-Hosted Runner Architecture"
sidebar_label: Runner Architecture
description: Decide how to create self-hosted runners
---

import Intro from "@site/src/components/Intro";
import Note from '@site/src/components/Note';
import Steps from "@site/src/components/Steps";
import Step from "@site/src/components/Step";
import StepNumber from "@site/src/components/StepNumber";

<Intro>

>   Self-hosted runners are custom runners that we use to run GitHub Actions workflows. We can create these runners a
>   number of ways.

Reviewer: ```suggestion
Decide on how to operate self-hosted runners that are used to run GitHub Actions workflows. These runners can be set up in various ways and allow us to avoid platform fees while running CI jobs in private infrastructure, enabling access to VPC resources. This approach is ideal for private repositories, providing control over instance size, architecture, and control costs by leveraging spot instances. The right choice depends on your platform, whether you’re using predominantly EKS, ECS, or Lambda.”
```

</Intro>

## Problem

> We support multiple options for deploying self-hosted runners. We can deploy runners with EKS, Philips Labs, or with an

Reviewer: When using GitHub Actions, you have the option of both GitHub Cloud-hosted and self-hosted runners, which can be used together. However, there are cases where self-hosted runners are essential. For example, if you’re using GitOps, private runners are needed to access resources inside a VPC, such as databases, Kubernetes API endpoints, and Kafka servers.

While self-hosted runners work well for private infrastructure, they’re not suitable for public or open-source repositories, as they can expose sensitive resources. If your organization contributes to open source, this must be considered when deciding on a runner strategy.

Your approach to self-hosted runners depends on your infrastructure. If you use Kubernetes, it’s often best to run your runners on Kubernetes. If you rely on ECS or Lambdas, you’ll want to avoid unnecessary Kubernetes dependencies and use a different hosting method for your runners.

For Kubernetes-based setups, proper node pool configuration with Karpenter is key to maintaining stability and ensuring effective auto-scaling with spot and on-demand instances. We offer several options for deploying self-hosted runners, including EKS, Philips Labs’ solution, or using an Auto Scaling Group (ASG). This decision focuses on the type, placement and management of the runners.

> We support multiple options for deploying self-hosted runners. We can deploy runners with EKS, Philips Labs, or with an
> ASG. For this decision, we will focus on the placement of the runners themselves.

Reviewer: ```suggestion
When using GitHub Actions, you have the option of both GitHub Cloud-hosted and self-hosted runners, which can be used together. However, there are cases where self-hosted runners are essential. For example, if you’re using GitOps, private runners are needed to access resources inside a VPC, such as databases, Kubernetes API endpoints, and Kafka servers.

While self-hosted runners work well for private infrastructure, they’re not suitable for public or open-source repositories, as they can expose sensitive resources. If your organization contributes to open source, this must be considered when deciding on a runner strategy.

Your approach to self-hosted runners depends on your infrastructure. If you use Kubernetes, it’s often best to run your runners on Kubernetes. If you rely on ECS or Lambdas, you’ll want to avoid unnecessary Kubernetes dependencies and use a different hosting method for your runners.

For Kubernetes-based setups, proper node pool configuration with Karpenter is key to maintaining stability and ensuring effective auto-scaling with spot and on-demand instances. We offer several options for deploying self-hosted runners, including EKS, Philips Labs’ solution, or using an Auto Scaling Group (ASG). This decision focuses on the type, placement and management of the runners.
```

### Option 3: GitHub Actions Runner on EKS (`eks/github-actions-runner`)

Alternatively, we can deploy the GitHub Actions Runner on EKS. This option is similar to the previous one, but it uses
the GitHub Actions Runner instead of the Actions Runner Controller.

This component deploys self-hosted GitHub Actions Runners and a
[Controller](https://docs.github.com/en/actions/hosting-your-own-runners/managing-self-hosted-runners-with-actions-runner-controller/quickstart-for-actions-runner-controller#introduction)
on an EKS cluster, using
"[runner scale sets](https://docs.github.com/en/actions/hosting-your-own-runners/managing-self-hosted-runners-with-actions-runner-controller/deploying-runner-scale-sets-with-actions-runner-controller#runner-scale-set)".

This solution is supported by GitHub and supersedes the
[actions-runner-controller](https://github.com/actions/actions-runner-controller/blob/master/docs/about-arc.md)
developed by Summerwind and deployed by Cloud Posse's
[actions-runner-controller](https://docs.cloudposse.com/components/library/aws/eks/actions-runner-controller/)
component.

However, there are some limitations to the official Runner Sets implementation:

> 

Reviewer: ```
### Option 4: Managed Runners

There are a number of services that offer managed runners. These still have the advantage over GitHub Cloud hosted runners as the can be managed within you private VPCs.

One option to consider is [runs-on.com](https://runs-on.com/) which provides a very inexpensive option.
```

> <Steps>

Reviewer: only use `<Steps/>` when they are truly steps.

For now, use this syntax:

```
- #### Limited Set of Packages
  The runner image used by Runner Sets contains [no more packages than are necessary](https://docs.github.com/en/actions/hosting-your-own-runners/managing-self-hosted-runners-with-actions-runner-controller/about-actions-runner-controller#about-the-runner-container-image) to run the runner. This is in contrast to the Summerwind implementation, which contains some commonly needed packages like `build-essential`, `curl`, `wget`, `git`, and `jq`, and the GitHub hosted images which contain a robust set of tools. (This is a limitation of the official Runner Sets implementation, not this component per se.) You will need to install any tools you need in your workflows, either as part of your workflow (recommended), by maintaining a [custom runner image](https://docs.github.com/en/actions/hosting-your-own-runners/managing-self-hosted-runners-with-actions-runner-controller/about-actions-runner-controller#creating-your-own-runner-image), or by running such steps in a [separate container](https://docs.github.com/en/actions/using-jobs/running-jobs-in-a-container) that has the tools pre-installed. Many tools have publicly ava...
  
- #### Docker in Docker (dind) mode only
  etc..
```

      runners.
    - All the runners use the same GitHub secret (app or PAT). Using a GitHub app is preferred anyway, and the single GitHub
      app serves the entire organization.
    - Only one controller is supported per cluster, though it can have multiple replicas.

  </Step>
</Steps>

These limitations could be addressed if there is demand. Contact [Cloud Posse Professional Services](https://cloudposse.com/professional-services/) if you would be interested in sponsoring the development of any of these features.

### Option 4: Philips Labs Runners (`philips-labs-github-runners`)

If we are not deploying EKS, it's not worth the additional effort to set up Self-Hosted runners on EKS. Instead, we deploy Self-Hosted runners on EC2 instances. These are managed by an API Gateway and Lambda function that will automatically scale the number of runners based on the number of pending jobs in the queue. The queue is written to by the API Gateway from GitHub Events.

For more on this option, see the [Philips Labs GitHub Runner](https://philips-labs.github.io/terraform-aws-github-runner/) documentation.

> ## Solution

Reviewer: ```suggestion
## Recommendation
```
